### PR TITLE
NoUselessElseFixer - fix wrong if handling

### PR DESCRIPTION
**Pull request by SpacePossum**
@ see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1923

### CODE REVIEW

**Comment by SpacePossum** [comment]
@localheinz sorry for messing this up :(
This is an early opening PR since when it tried to find the problem based on your findings I ran into an `out range exception`, so not a false-positive fix. I've to dig a bit more to see if there is another cause.

If you have the time, could you retest the fixer with the changes in this PR on your code that was fixed incorrectly?

File: src/Fixer/ControlStructure/NoUselessElseFixer.php [hunk 1]
```diff
@@ -100,21 +100,24 @@ private function fixElse(Tokens $tokens, $index)
...
+                    // positives
                     array(T_BREAK),
                     array(T_CONTINUE),
                     array(T_EXIT),
                     array(T_GOTO),
                     array(T_RETURN),
                     array(T_THROW),
+                    // negatives
+                    ';',
+                    array(T_IF),
```
Comment (SpacePossum) [inline src/Fixer/ControlStructure/NoUselessElseFixer.php]: What would be even better/needed is to have a limit for the `getTokenOfKindSibling`method.

File: tests/Fixer/ControlStructure/NoUselessElseFixerTest.php [hunk 2]
```diff
@@ -22,6 +22,21 @@
  */
 final class NoUselessElseFixerTest extends AbstractFixerTestCase
 {
+    public function testLocalheinz()
+    {
+        $input =
+            '<?php
+                return $tmp(function () use $a {
```
Comment (localheinz) [inline tests/Fixer/ControlStructure/NoUselessElseFixerTest.php]: @SpacePossum 

Should be `use ($a)`!

😉

File: tests/Fixer/ControlStructure/NoUselessElseFixerTest.php [hunk 3]
```diff
@@ -22,6 +22,21 @@
  */
 final class NoUselessElseFixerTest extends AbstractFixerTestCase
 {
+    public function testLocalheinz()
```
Comment (keradus) [inline tests/Fixer/ControlStructure/NoUselessElseFixerTest.php]: ...

File: tests/Fixer/ControlStructure/NoUselessElseFixerTest.php [hunk 4]
```diff
@@ -22,6 +22,21 @@
...
+    {
+        $input =
+            '<?php
+                return $tmp(function () use $a {
+                    if ($a) {
+                        $a++;
+                    } else {
+                        $a--;
+                    }
+                };
```
Comment (localheinz) [inline tests/Fixer/ControlStructure/NoUselessElseFixerTest.php]: @SpacePossum 

You are also missing a closing `)` here!

Comment (SpacePossum) [inline tests/Fixer/ControlStructure/NoUselessElseFixerTest.php]: thanks! I'm re-doing this whole fix ATM, what I need is a limit on the `getTokenOfKindSibling` method. Hope to fix this evening!

Comment (localheinz) [inline tests/Fixer/ControlStructure/NoUselessElseFixerTest.php]: @SpacePossum 

Cool, let me know when I should take a look and test!

Comment (GrahamCampbell) [inline tests/Fixer/ControlStructure/NoUselessElseFixerTest.php]: :+1:

File: src/Tokenizer/Tokens.php [hunk 5]
```diff
@@ -593,39 +594,50 @@ public function getPrevNonWhitespace($index, $whitespaces = null)
...
+     * @param int      $index         token index
+     * @param int      $direction     direction for looking, +1 or -1
+     * @param int|null $limit         null no limit
+     * @param array    $tokens        possible tokens
+     * @param bool     $caseSensitive perform a case sensitive comparison
      *
      * @return int|null
      */
-    public function getTokenOfKindSibling($index, $direction, array $tokens = array(), $caseSensitive = true)
+    public function getTokenOfKindSibling($index, $direction, $limit = null, array $tokens = array(), $caseSensitive = true)
```
Comment (SpacePossum) [inline src/Tokenizer/Tokens.php]: This is the only real point of interest in this PR.

Comment (keradus) [inline src/Tokenizer/Tokens.php]: why do we need it in the first place? what was wrong with checking it on the calling place?
it's not common in our code to check that limit as far as I can see

File: src/Tokenizer/Tokens.php [hunk 6]
```diff
@@ -593,39 +594,50 @@ public function getPrevNonWhitespace($index, $whitespaces = null)
...
+     * @param bool     $caseSensitive perform a case sensitive comparison
+     * @param int|null $limit         false no limit
      *
      * @return int|null
      */
-    public function getPrevTokenOfKind($index, array $tokens = array(), $caseSensitive = true)
+    public function getPrevTokenOfKind($index, array $tokens = array(), $caseSensitive = true, $limit = null)
     {
-        return $this->getTokenOfKindSibling($index, -1, $tokens, $caseSensitive);
+        return $this->getTokenOfKindSibling($index, -1, $limit, $tokens, $caseSensitive);
```
Comment (keradus) [inline src/Tokenizer/Tokens.php]: also, having different order of params for general `getTokenOfKindSibling` method and it's aliases is not intuitive

**Comment by keradus** [comment]
A lot of nice celanup, but thanks to them I lost what part is actually a bugfix and what only cleanup

**Comment by Soullivaneuh** [comment]
@SpacePossum Could you add this to your tests? https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1923#issuecomment-222731341

Thanks.

**Comment by SpacePossum** [comment]
sure thing @Soullivaneuh , sorry for the inconvenience (cc @greg0ire) :(
while we are talking, could take a look at https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1942 IIRC your are the author of the fixer?

**Comment by SpacePossum** [comment]
ready for review

**Comment by keradus** [comment]
:+1:

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Pull request merged**